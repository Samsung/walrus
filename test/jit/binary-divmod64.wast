(module
(func (export "test1") (result i64 i64 i64 i64 i64) (local i64 i64 i64)
  i64.const 9876543210
  local.set 0
  i64.const 123456789123456
  local.set 1
  i64.const -1
  local.set 2

  i64.const 1234567812345678
  local.get 0
  i64.div_s

  local.get 1
  i64.const 567567567
  i64.rem_s

  i64.const -34567893456789
  local.get 0
  i64.div_s

  local.get 1
  local.get 0
  i64.rem_s

  i64.const -0x7fffffffffffffff
  local.get 2
  i64.div_s

  (; 124999, 59517183, 18446744073709548117, 9875541666, 9223372036854775807 ;)
)

(func (export "test2") (result i64 i64 i64 i64 i64) (local i64 i64 i64)
  i64.const 7654376543
  local.set 0
  i64.const 23456782345678
  local.set 1
  i64.const 1
  local.set 2

  i64.const 1234567812345678
  local.get 0
  i64.div_u

  local.get 1
  i64.const 87654321
  i64.rem_u

  i64.const 0xff00000000000000
  local.get 0
  i64.div_u

  local.get 1
  local.get 0
  i64.rem_u

  i64.const 0xffffffffffffffff
  local.get 2
  i64.div_s

  (; 161289, 47774473, 2400546455, 3772617926, 18446744073709551615 ;)
)

(func (export "test3") (result i64) (local i64)
  i64.const 0
  local.set 0

  i64.const 5
  local.get 0
  i64.div_s
  (; exception ;)
)

(func (export "test4") (result i64) (local i64)
  i64.const -1
  local.set 0

  i64.const 0x8000000000000000
  local.get 0
  i64.div_s
  (; exception ;)
)

(func (export "test5") (result i64) (local i64)
  i64.const 0
  local.set 0

  i64.const 5
  local.get 0
  i64.rem_s
  (; exception ;)
)

(func (export "test6") (result i64) (local i64)
  i64.const -1
  local.set 0

  i64.const 0x8000000000000000
  local.get 0
  i64.rem_s
  (; exception ;)
)

(func (export "test7") (result i64) (local i64)
  i64.const 0
  local.set 0

  i64.const 5
  local.get 0
  i64.div_u
  (; exception ;)
)

(func (export "test8") (result i64) (local i64)
  i64.const 0
  local.set 0

  i64.const 5
  local.get 0
  i64.rem_u
  (; exception ;)
)
)

(assert_return (invoke "test1") (i64.const 124999) (i64.const 59517183) (i64.const 18446744073709548117) (i64.const 9875541666) (i64.const 9223372036854775807))
(assert_return (invoke "test2") (i64.const 161289) (i64.const 47774473) (i64.const 2400546455) (i64.const 3772617926) (i64.const 18446744073709551615))
(assert_trap (invoke "test3") "integer divide by zero")
(assert_trap (invoke "test5") "integer divide by zero")
(assert_trap (invoke "test7") "integer divide by zero")
(assert_trap (invoke "test8") "integer divide by zero")

