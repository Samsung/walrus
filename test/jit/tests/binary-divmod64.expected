[[[[[[[  Function   0  ]]]]]]]
Local 0: mapped to: 0 range [1 - 16]
Local 1: mapped to: 1 range [3 - 15]
Local 2: mapped to: 2 range [5 - 19]
Number of locals reduced from 3 to 3
0: Opcode: i64.const
  Result[0]: i64
1: Opcode: local.set
  Index: 0
  Param[0]: ref:0(0)
2: Opcode: i64.const
  Result[0]: i64
3: Opcode: local.set
  Index: 1
  Param[0]: ref:2(0)
4: Opcode: i64.const
  Result[0]: i64
5: Opcode: local.set
  Index: 2
  Param[0]: ref:4(0)
6: Opcode: i64.const
  Result[0]: i64
7: Opcode: local.get
  Index: 0
  Result[0]: i64
8: Opcode: i64.div_s
  Param[0]: ref:6(0)
  Param[1]: ref:7(0)
  Result[0]: i64
9: Opcode: local.get
  Index: 1
  Result[0]: i64
10: Opcode: i64.const
  Result[0]: i64
11: Opcode: i64.rem_s
  Param[0]: ref:9(0)
  Param[1]: ref:10(0)
  Result[0]: i64
12: Opcode: i64.const
  Result[0]: i64
13: Opcode: local.get
  Index: 0
  Result[0]: i64
14: Opcode: i64.div_s
  Param[0]: ref:12(0)
  Param[1]: ref:13(0)
  Result[0]: i64
15: Opcode: local.get
  Index: 1
  Result[0]: i64
16: Opcode: local.get
  Index: 0
  Result[0]: i64
17: Opcode: i64.rem_s
  Param[0]: ref:15(0)
  Param[1]: ref:16(0)
  Result[0]: i64
18: Opcode: i64.const
  Result[0]: i64
19: Opcode: local.get
  Index: 2
  Result[0]: i64
20: Opcode: i64.div_s
  Param[0]: ref:18(0)
  Param[1]: ref:19(0)
  Result[0]: i64
21: Opcode: return
  Param[0]: ref:8(0)
  Param[1]: ref:11(0)
  Param[2]: ref:14(0)
  Param[3]: ref:17(0)
  Param[4]: ref:20(0)
------------------------------
Frame size: 0, Arguments size: 64 (total: 64)
0: Opcode: i64.const
  Result[0]: i64 unused
1: Opcode: local.set
  Addr: sp[40]
  Param[0]: i64 const:0
2: Opcode: i64.const
  Result[0]: i64 unused
3: Opcode: local.set
  Addr: sp[48]
  Param[0]: i64 const:2
4: Opcode: i64.const
  Result[0]: i64 unused
5: Opcode: local.set
  Addr: sp[56]
  Param[0]: i64 const:4
6: Opcode: i64.const
  Result[0]: i64 unused
7: Opcode: local.get
  Addr: sp[40]
  Result[0]: i64 unused
8: Opcode: i64.div_s
  Param[0]: i64 const:6
  Param[1]: i64 sp[40]
  Result[0]: i64 sp[0]
9: Opcode: local.get
  Addr: sp[48]
  Result[0]: i64 unused
10: Opcode: i64.const
  Result[0]: i64 unused
11: Opcode: i64.rem_s
  Param[0]: i64 sp[48]
  Param[1]: i64 const:10
  Result[0]: i64 sp[8]
12: Opcode: i64.const
  Result[0]: i64 unused
13: Opcode: local.get
  Addr: sp[40]
  Result[0]: i64 unused
14: Opcode: i64.div_s
  Param[0]: i64 const:12
  Param[1]: i64 sp[40]
  Result[0]: i64 sp[16]
15: Opcode: local.get
  Addr: sp[48]
  Result[0]: i64 unused
16: Opcode: local.get
  Addr: sp[40]
  Result[0]: i64 unused
17: Opcode: i64.rem_s
  Param[0]: i64 sp[48]
  Param[1]: i64 sp[40]
  Result[0]: i64 sp[24]
18: Opcode: i64.const
  Result[0]: i64 unused
19: Opcode: local.get
  Addr: sp[56]
  Result[0]: i64 unused
20: Opcode: i64.div_s
  Param[0]: i64 const:18
  Param[1]: i64 sp[56]
  Result[0]: i64 sp[32]
21: Opcode: return
  Param[0]: i64 sp[0]
  Param[1]: i64 sp[8]
  Param[2]: i64 sp[16]
  Param[3]: i64 sp[24]
  Param[4]: i64 sp[32]

[[[[[[[  Function   1  ]]]]]]]
Local 0: mapped to: 0 range [1 - 16]
Local 1: mapped to: 1 range [3 - 15]
Local 2: mapped to: 2 range [5 - 19]
Number of locals reduced from 3 to 3
0: Opcode: i64.const
  Result[0]: i64
1: Opcode: local.set
  Index: 0
  Param[0]: ref:0(0)
2: Opcode: i64.const
  Result[0]: i64
3: Opcode: local.set
  Index: 1
  Param[0]: ref:2(0)
4: Opcode: i64.const
  Result[0]: i64
5: Opcode: local.set
  Index: 2
  Param[0]: ref:4(0)
6: Opcode: i64.const
  Result[0]: i64
7: Opcode: local.get
  Index: 0
  Result[0]: i64
8: Opcode: i64.div_u
  Param[0]: ref:6(0)
  Param[1]: ref:7(0)
  Result[0]: i64
9: Opcode: local.get
  Index: 1
  Result[0]: i64
10: Opcode: i64.const
  Result[0]: i64
11: Opcode: i64.rem_u
  Param[0]: ref:9(0)
  Param[1]: ref:10(0)
  Result[0]: i64
12: Opcode: i64.const
  Result[0]: i64
13: Opcode: local.get
  Index: 0
  Result[0]: i64
14: Opcode: i64.div_u
  Param[0]: ref:12(0)
  Param[1]: ref:13(0)
  Result[0]: i64
15: Opcode: local.get
  Index: 1
  Result[0]: i64
16: Opcode: local.get
  Index: 0
  Result[0]: i64
17: Opcode: i64.rem_u
  Param[0]: ref:15(0)
  Param[1]: ref:16(0)
  Result[0]: i64
18: Opcode: i64.const
  Result[0]: i64
19: Opcode: local.get
  Index: 2
  Result[0]: i64
20: Opcode: i64.div_s
  Param[0]: ref:18(0)
  Param[1]: ref:19(0)
  Result[0]: i64
21: Opcode: return
  Param[0]: ref:8(0)
  Param[1]: ref:11(0)
  Param[2]: ref:14(0)
  Param[3]: ref:17(0)
  Param[4]: ref:20(0)
------------------------------
Frame size: 0, Arguments size: 64 (total: 64)
0: Opcode: i64.const
  Result[0]: i64 unused
1: Opcode: local.set
  Addr: sp[40]
  Param[0]: i64 const:0
2: Opcode: i64.const
  Result[0]: i64 unused
3: Opcode: local.set
  Addr: sp[48]
  Param[0]: i64 const:2
4: Opcode: i64.const
  Result[0]: i64 unused
5: Opcode: local.set
  Addr: sp[56]
  Param[0]: i64 const:4
6: Opcode: i64.const
  Result[0]: i64 unused
7: Opcode: local.get
  Addr: sp[40]
  Result[0]: i64 unused
8: Opcode: i64.div_u
  Param[0]: i64 const:6
  Param[1]: i64 sp[40]
  Result[0]: i64 sp[0]
9: Opcode: local.get
  Addr: sp[48]
  Result[0]: i64 unused
10: Opcode: i64.const
  Result[0]: i64 unused
11: Opcode: i64.rem_u
  Param[0]: i64 sp[48]
  Param[1]: i64 const:10
  Result[0]: i64 sp[8]
12: Opcode: i64.const
  Result[0]: i64 unused
13: Opcode: local.get
  Addr: sp[40]
  Result[0]: i64 unused
14: Opcode: i64.div_u
  Param[0]: i64 const:12
  Param[1]: i64 sp[40]
  Result[0]: i64 sp[16]
15: Opcode: local.get
  Addr: sp[48]
  Result[0]: i64 unused
16: Opcode: local.get
  Addr: sp[40]
  Result[0]: i64 unused
17: Opcode: i64.rem_u
  Param[0]: i64 sp[48]
  Param[1]: i64 sp[40]
  Result[0]: i64 sp[24]
18: Opcode: i64.const
  Result[0]: i64 unused
19: Opcode: local.get
  Addr: sp[56]
  Result[0]: i64 unused
20: Opcode: i64.div_s
  Param[0]: i64 const:18
  Param[1]: i64 sp[56]
  Result[0]: i64 sp[32]
21: Opcode: return
  Param[0]: i64 sp[0]
  Param[1]: i64 sp[8]
  Param[2]: i64 sp[16]
  Param[3]: i64 sp[24]
  Param[4]: i64 sp[32]

[[[[[[[  Function   2  ]]]]]]]
Local 0: mapped to: 0 range [1 - 3]
Number of locals reduced from 1 to 1
0: Opcode: i64.const
  Result[0]: i64
1: Opcode: local.set
  Index: 0
  Param[0]: ref:0(0)
2: Opcode: i64.const
  Result[0]: i64
3: Opcode: local.get
  Index: 0
  Result[0]: i64
4: Opcode: i64.div_s
  Param[0]: ref:2(0)
  Param[1]: ref:3(0)
  Result[0]: i64
5: Opcode: return
  Param[0]: ref:4(0)
------------------------------
Frame size: 0, Arguments size: 16 (total: 16)
0: Opcode: i64.const
  Result[0]: i64 unused
1: Opcode: local.set
  Addr: sp[8]
  Param[0]: i64 const:0
2: Opcode: i64.const
  Result[0]: i64 unused
3: Opcode: local.get
  Addr: sp[8]
  Result[0]: i64 unused
4: Opcode: i64.div_s
  Param[0]: i64 const:2
  Param[1]: i64 sp[8]
  Result[0]: i64 sp[0]
5: Opcode: return
  Param[0]: i64 sp[0]

[[[[[[[  Function   3  ]]]]]]]
Local 0: mapped to: 0 range [1 - 3]
Number of locals reduced from 1 to 1
0: Opcode: i64.const
  Result[0]: i64
1: Opcode: local.set
  Index: 0
  Param[0]: ref:0(0)
2: Opcode: i64.const
  Result[0]: i64
3: Opcode: local.get
  Index: 0
  Result[0]: i64
4: Opcode: i64.div_s
  Param[0]: ref:2(0)
  Param[1]: ref:3(0)
  Result[0]: i64
5: Opcode: return
  Param[0]: ref:4(0)
------------------------------
Frame size: 0, Arguments size: 16 (total: 16)
0: Opcode: i64.const
  Result[0]: i64 unused
1: Opcode: local.set
  Addr: sp[8]
  Param[0]: i64 const:0
2: Opcode: i64.const
  Result[0]: i64 unused
3: Opcode: local.get
  Addr: sp[8]
  Result[0]: i64 unused
4: Opcode: i64.div_s
  Param[0]: i64 const:2
  Param[1]: i64 sp[8]
  Result[0]: i64 sp[0]
5: Opcode: return
  Param[0]: i64 sp[0]

[[[[[[[  Function   4  ]]]]]]]
Local 0: mapped to: 0 range [1 - 3]
Number of locals reduced from 1 to 1
0: Opcode: i64.const
  Result[0]: i64
1: Opcode: local.set
  Index: 0
  Param[0]: ref:0(0)
2: Opcode: i64.const
  Result[0]: i64
3: Opcode: local.get
  Index: 0
  Result[0]: i64
4: Opcode: i64.rem_s
  Param[0]: ref:2(0)
  Param[1]: ref:3(0)
  Result[0]: i64
5: Opcode: return
  Param[0]: ref:4(0)
------------------------------
Frame size: 0, Arguments size: 16 (total: 16)
0: Opcode: i64.const
  Result[0]: i64 unused
1: Opcode: local.set
  Addr: sp[8]
  Param[0]: i64 const:0
2: Opcode: i64.const
  Result[0]: i64 unused
3: Opcode: local.get
  Addr: sp[8]
  Result[0]: i64 unused
4: Opcode: i64.rem_s
  Param[0]: i64 const:2
  Param[1]: i64 sp[8]
  Result[0]: i64 sp[0]
5: Opcode: return
  Param[0]: i64 sp[0]

[[[[[[[  Function   5  ]]]]]]]
Local 0: mapped to: 0 range [1 - 3]
Number of locals reduced from 1 to 1
0: Opcode: i64.const
  Result[0]: i64
1: Opcode: local.set
  Index: 0
  Param[0]: ref:0(0)
2: Opcode: i64.const
  Result[0]: i64
3: Opcode: local.get
  Index: 0
  Result[0]: i64
4: Opcode: i64.rem_s
  Param[0]: ref:2(0)
  Param[1]: ref:3(0)
  Result[0]: i64
5: Opcode: return
  Param[0]: ref:4(0)
------------------------------
Frame size: 0, Arguments size: 16 (total: 16)
0: Opcode: i64.const
  Result[0]: i64 unused
1: Opcode: local.set
  Addr: sp[8]
  Param[0]: i64 const:0
2: Opcode: i64.const
  Result[0]: i64 unused
3: Opcode: local.get
  Addr: sp[8]
  Result[0]: i64 unused
4: Opcode: i64.rem_s
  Param[0]: i64 const:2
  Param[1]: i64 sp[8]
  Result[0]: i64 sp[0]
5: Opcode: return
  Param[0]: i64 sp[0]

[[[[[[[  Function   6  ]]]]]]]
Local 0: mapped to: 0 range [1 - 3]
Number of locals reduced from 1 to 1
0: Opcode: i64.const
  Result[0]: i64
1: Opcode: local.set
  Index: 0
  Param[0]: ref:0(0)
2: Opcode: i64.const
  Result[0]: i64
3: Opcode: local.get
  Index: 0
  Result[0]: i64
4: Opcode: i64.div_u
  Param[0]: ref:2(0)
  Param[1]: ref:3(0)
  Result[0]: i64
5: Opcode: return
  Param[0]: ref:4(0)
------------------------------
Frame size: 0, Arguments size: 16 (total: 16)
0: Opcode: i64.const
  Result[0]: i64 unused
1: Opcode: local.set
  Addr: sp[8]
  Param[0]: i64 const:0
2: Opcode: i64.const
  Result[0]: i64 unused
3: Opcode: local.get
  Addr: sp[8]
  Result[0]: i64 unused
4: Opcode: i64.div_u
  Param[0]: i64 const:2
  Param[1]: i64 sp[8]
  Result[0]: i64 sp[0]
5: Opcode: return
  Param[0]: i64 sp[0]

[[[[[[[  Function   7  ]]]]]]]
Local 0: mapped to: 0 range [1 - 3]
Number of locals reduced from 1 to 1
0: Opcode: i64.const
  Result[0]: i64
1: Opcode: local.set
  Index: 0
  Param[0]: ref:0(0)
2: Opcode: i64.const
  Result[0]: i64
3: Opcode: local.get
  Index: 0
  Result[0]: i64
4: Opcode: i64.rem_u
  Param[0]: ref:2(0)
  Param[1]: ref:3(0)
  Result[0]: i64
5: Opcode: return
  Param[0]: ref:4(0)
------------------------------
Frame size: 0, Arguments size: 16 (total: 16)
0: Opcode: i64.const
  Result[0]: i64 unused
1: Opcode: local.set
  Addr: sp[8]
  Param[0]: i64 const:0
2: Opcode: i64.const
  Result[0]: i64 unused
3: Opcode: local.get
  Addr: sp[8]
  Result[0]: i64 unused
4: Opcode: i64.rem_u
  Param[0]: i64 const:2
  Param[1]: i64 sp[8]
  Result[0]: i64 sp[0]
5: Opcode: return
  Param[0]: i64 sp[0]

test1() => i64:124999, i64:59517183, i64:18446744073709548117, i64:9875541666, i64:9223372036854775807
test2() => i64:161289, i64:47774473, i64:2400546455, i64:3772617926, i64:18446744073709551615
test3() => i64:0
test4() => i64:0
test5() => i64:0
test6() => i64:0
test7() => i64:0
test8() => i64:0
